File: 3_fractional_knapsack.py
def fractional_knapsack():
    # Step 1: Take number of items
    n = int(input("Enter number of items: "))

    weights = []
    values = []

    # Step 2: Take weight and value for each item
    print("\nEnter weight and value for each item:")
    for i in range(n):
        w = float(input(f"Weight of item {i+1}: "))
        v = float(input(f"Value of item {i+1}: "))
        weights.append(w)
        values.append(v)

    # Step 3: Take knapsack capacity
    capacity = float(input("\nEnter knapsack capacity: "))

    # Step 4: Calculate value/weight ratio for each item
    ratio = []
    for i in range(n):
        ratio.append(values[i] / weights[i])

    # Step 5: Sort items manually by ratio (descending)
    for i in range(n):
        for j in range(i + 1, n):
            if ratio[i] < ratio[j]:
                # Swap ratios
                ratio[i], ratio[j] = ratio[j], ratio[i]
                # Swap weights
                weights[i], weights[j] = weights[j], weights[i]
                # Swap values
                values[i], values[j] = values[j], values[i]

    # Step 6: Apply greedy selection
    total_value = 0.0
    print("\nItem selection process:")
    for i in range(n):
        if capacity == 0:
            break
        if weights[i] <= capacity:
            total_value += values[i]
            capacity -= weights[i]
            print(f"  Took full item (weight={weights[i]}, value={values[i]})")
        else:
            fraction = capacity / weights[i]
            total_value += values[i] * fraction
            print(f"  Took {fraction*100:.1f}% of item (weight={weights[i]}, value={values[i]})")
            capacity = 0

    print(f"\nMaximum value in knapsack = {total_value:.2f}")


# Run the program
if __name__ == "__main__":
    fractional_knapsack()


# Enter number of items: 3
# Enter weight and value for each item:
# Weight of item 1: 10
# Value of item 1: 60
# Weight of item 2: 20
# Value of item 2: 100
# Weight of item 3: 30
# Value of item 3: 120

# Enter knapsack capacity: 50

File: 3_fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <iomanip>
using namespace std;

void fractionalKnapsack() {
    // Step 1: Take number of items
    int n;
    cout << "Enter number of items: ";
    cin >> n;

    vector<double> weights(n), values(n);

    // Step 2: Take weight and value for each item
    cout << "\nEnter weight and value for each item:\n";
    for (int i = 0; i < n; i++) {
        cout << "Weight of item " << i + 1 << ": ";
        cin >> weights[i];
        cout << "Value of item " << i + 1 << ": ";
        cin >> values[i];
    }

    // Step 3: Take knapsack capacity
    double capacity;
    cout << "\nEnter knapsack capacity: ";
    cin >> capacity;

    // Step 4: Calculate value/weight ratio for each item
    vector<double> ratio(n);
    for (int i = 0; i < n; i++) {
        ratio[i] = values[i] / weights[i];
    }

    // Step 5: Sort items manually by ratio (descending)
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (ratio[i] < ratio[j]) {
                swap(ratio[i], ratio[j]);
                swap(weights[i], weights[j]);
                swap(values[i], values[j]);
            }
        }
    }

    // Step 6: Apply greedy selection
    double totalValue = 0.0;
    cout << fixed << setprecision(2);
    cout << "\nItem selection process:\n";

    for (int i = 0; i < n; i++) {
        if (capacity == 0)
            break;

        if (weights[i] <= capacity) {
            totalValue += values[i];
            capacity -= weights[i];
            cout << "  Took full item (weight=" << weights[i]
                 << ", value=" << values[i] << ")\n";
        } else {
            double fraction = capacity / weights[i];
            totalValue += values[i] * fraction;
            cout << "  Took " << fraction * 100 << "% of item (weight=" << weights[i]
                 << ", value=" << values[i] << ")\n";
            capacity = 0;
        }
    }

    cout << "\nMaximum value in knapsack = " << totalValue << endl;
}

// Step 7: Run the program
int main() {
    fractionalKnapsack();
    return 0;
}
